### DISCUSSION TITLE
Dynamic number of grid columns based on number of children elements, in a purge-friendly way?

### DISCUSSION THREAD

**Discussion by raphaelbernhart** [tailwindlabs/tailwindcss#5577] [Help] [answered]
Hello there,

I am currently facing an error with tailwindcss compiling. I am having a class which is not compiled to css.

My techstack is Vue3 TS + Tailwindcss + PostCSS.

I am having a table component with the following markup
```
<div :class="`w-full grid grid-cols-2 md:grid-cols-${headings.length} gap-4 p-3`">
    <span v-for="(heading, index) in headings" :key="heading" class="font-thin text-sm text-gray-500" :class="{ 'hidden md:table-cell': showCell(index) }">
                {{ heading }}
    </span>
</div>
```

The classes `grid` and `grid-cols-2` are working fine after vue build but the class md:grid-cols-${any number} is compiled proper from vue (template syntax)
<img width="412" alt="Bildschirmfoto 2021-09-23 um 08 36 25" src="https://user-images.githubusercontent.com/48283236/134464160-b38da341-41e4-4f91-b36b-f931b0ec7103.png">
but the minified css file isn't compiled correctly. The class is just not compiled at all in the build.
When I search for the class in the minified css in the dist folder it is just not there
<img width="1200" alt="Bildschirmfoto 2021-09-23 um 08 43 31" src="https://user-images.githubusercontent.com/48283236/134464650-bd084c0a-51d9-456c-8a08-5b064c2295e2.png">

I think this has something to do with the compilation with the vue syntax of dynamic markup?
Maybe PostCSS?

Currently workaround this with just custom defining all css classes. But this just isn't good.

Thanks anyways for the help :)

**Comment by simonswiss** [comment, marked as answer]
Hey!

**TL;DR: you could possibly achieve what you want with CSS Grid!**

---

Neither of PurgeCSS or Tailwind `JIT` will evaluate your code when looking for classes – they both just look for complete text strings.

When you write `grid-cols-${headings.length}`, none of the `grid-cols-*` utilities are recognised, and therefore none of them make the final "cut" in the generated CSS.

You can read more about it on this page of the Tailwind CSS documentation:

https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html

As you mentioned, a solution is to create a "lookup table" object that lists all the possible options. In some situations, it makes sense:

```js
const colorsLookup = {
primary: 'bg-green-800 text-green-100',
secondary: 'bg-indigo-800 text-indigo-100',
// ...
}
```

... and then you could use some prop/state value to use the classes:

```<button :class="colorsLookup[theme]">```

But if in your scenario, you could have from 1 to 300 headings, this wouldn't be practical..

---

# CSS Grid instead?

May suggest a totally different approach here? Instead of counting the number of headings with JS and enforcing a specific amount of grid columns that way, you can use the `auto-fit` property of CSS Grid, to automatically split the content in the right amount of columns 👍

Paired with `minmax`, it allows you to define what minimum width a single column should have, and break to a second row if there is not enough space.

The following CSS will make each grid child item use it's own "share" of the available space, until there is not enough space left to respect the minimum of 200px, in which case it will break into two lines:

```css
repeat(auto-fit, minmax(200px, 1fr))
```

https://user-images.githubusercontent.com/485747/134612958-2410502e-e277-463e-a625-42f250c71f78.mov

If you don't care about a minimum width, you can use `minmax(0, 1fr)` and the items will _always_ stay on one row:

https://user-images.githubusercontent.com/485747/134613041-fb2fee25-256e-419e-92e1-f133d8176617.mov

Here's a Tailwind Play demo (where the screen recordings came from!):

https://play.tailwindcss.com/w9oTbP5AWW

Be sure to check the config file to see how I extended the `gridTemplateColumns` object to create these `auto-fit` utility classes 👍

Hope it helps!

**Reply by raphaelbernhart** [reply]
First of all thanks for the fast reply! This really helped me understanding the underlying concept of PurgeCSS.

As I am using css grid not very often I haven't tought about `auto-fit`, which is even a better solution for this problem than mine.

**Reply by simonswiss** [reply]
Awesome – glad to hear it was helpful ✨
